File: LeetCode/1047.py
#Link: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/

class Solution:
    def removeDuplicates(self, S: str) -> str:
        l, r = 0, 1
        while r < len(S):
            if S[l] == S[r]:
                S = S[:l] + S[r+1:]
                if l > 0 : l -= 1
                r = l + 1
            else:
                l = r
                r += 1
        return S

File: LeetCode/1047.cpp
//Link: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/

class Solution {
public:
    string removeDuplicates(string S) {
        int l = 0, r = 1;
        string result = "";
        while(r < S.length()) {
            if(S[l] == S[r]) {
                S[l] = S[r] = '0';
                while(l > 0 && S[l] == '0')
                    l--;
                r++;
            }
            else {
                l = r;
                r++;
            }
        }
        for(char ch : S)
            if(ch != '0')
                result += ch;
        return result;
    }
};
